# Step 1: 数値や文字の扱いを理解しよう。
- 1-1. 変数とデータ型
- 1-2. 関数と引数
- 1-3. 型変換と入力
- 1-4. print()による出力とf-string整形
- 1-5. コメントアウトとエラーメッセージ
- 1-6. 算術演算子と優先順位
- 1-7. import文と標準ライブラリ

# Step 1 の学習目標
- 変数とは何かを理解する
- 数値や文字列を使ってデータを表現できるようにする
- コメントアウトやエラーメッセージの構文を理解する

# 1-1. 変数とデータ型

## 変数とは？
- プログラム中で値（データ）を一時的に保存したり、計算に利用するために「変数（variable）」を使う。  
変数は「名前のついたデータの入れ物」と考えることができる。
- 変数に値を格納するには =（代入演算子）を使う。
- 以下ではtemperatureという変数に23.8という浮動小数点型の値を、locationという変数に文字列を、それぞれ格納する。

In [ ]:
temperature = 23.8   # 水温（℃）
location = "Lake Biwa"  # 地点名

## 変数のデータ型

- 変数には種類（データ型, type）がある。
- 水温なら小数を扱う**浮動小数点数（float）**、地点名なら文字列（string, `str`）。
- Pythonでは型を自動で判断してくれるため、C言語やJavaのように `int x;` のような宣言は不要。

---

## 型の例

| データの種類 | 型の名前 | 例 |
|---------------|------------|----|
| 整数 | `int` | 10, -5, 2025 |
| 実数（小数） | `float` | 23.8, -1.2, 3.14 |
| 文字列 | `str` | "Lake Biwa", 'Tokyo' |
| 論理値 | `bool` | True, False |

※ Pythonでは文字列を表すのにシングルクオテーション('')とダブルクオテーション("")の両方を用いる。1つのプロジェクトの中では一貫してどちらかを使うことを推奨。

## 型を調べる：`type()` 関数

Python では `type(変数名)` を使うと、その変数の型を調べることができます。

In [2]:
temperature = 23.8
location = "Lake Biwa"
print(type(temperature))
print(type(location))

<class 'float'>
<class 'str'>


In [ ]:
print(location + "の水温は" + str(temperature) + "℃です。")

# 1-2. 関数と引数

- 「何かの処理をまとめた部品」を **関数（function）** と呼ぶ。
- print() や type() はPythonに最初から用意されている関数。
- 関数を使うことで、同じ処理を何度でも簡単に実行できる。

In [ ]:
print("Hello")
print("Hello World")

- print：関数の名前
- ()（かっこ）：関数を実行する合図
- (arg)：関数に渡す情報（これを 引数 argument という）

- 引数はカンマで区切って複数渡すことができる。以下はprint関数の引数をカンマで区切って渡した場合。

In [5]:
print("気温は", 23.5, "度です。")

気温は 23.5 度です。


- print関数は文字列を + で結合して表示することもできる。以下改めて紹介するため、上記の書式と比較してみよう。

# 1-3. 型変換とキーボード入力

## 型変換
- Pythonでは、ある型の値を別の型に変換することができる。  
- これを **型の変換（type conversion）** または **キャスト（casting）** と呼ぶ。
- 例:


In [ ]:
int("5")    # 文字列 "5" を整数 5 に変換
float("3.14")  # 文字列 "3.14" を小数 3.14 に変換
str(10)     # 整数 10 を文字列 "10" に変換

| 変換先の型 | 関数        | 説明                    | 例                                   |
| ----- | --------- | --------------------- | ----------------------------------- |
| 整数    | `int()`   | 小数や数値文字列を整数に変換        | `int("10") → 10`                    |
| 小数    | `float()` | 整数や数値文字列を小数に変換        | `float("3.5") → 3.5`                |
| 文字列   | `str()`   | 数値や論理値を文字列に変換         | `str(25) → "25"`                    |
| 論理値   | `bool()`  | 値を真偽（True / False）に変換 | `bool(0) → False`, `bool(1) → True` |


## キーボードからの入力
- input()という関数は、キーボードから入力された文字を文字列(str)として返す。
- キーボードから入力した値（input()）は文字列（str型）として扱われるため、計算に使うには数値型（intやfloat）へ変換する必要がある。

In [8]:
print(input("入力された文字列を表示します:"))

入力された文字列を表示します: Python


Python


In [13]:
strin = input("摂氏の温度を入力してください:")
print(float(strin) + 273.15)

摂氏の温度を入力してください: 15


288.15


# 1-4. print()による出力とf-string整形

## 出力整形
- 出力整形とは、数値や文字列の見た目をきれいに整えることである。
- 見た目を整えることは一見「本質的ではない」ように思えるかもしれないが、実験結果や観測値を確認するのは人間である。人が視覚的に理解しやすい形でデータを出力できるようにすることには大きな意義がある。
- print()の場合、「小数点以下2桁まで表示したい」といったときf-string整形を用いる。

## f-string整形とは？
- f-string（フォーマット文字列）とは、文字列の前に f を付けて、中かっこ {} の中に変数や式を直接書ける構文。
- {} の中では 小数点以下の桁数や幅などを指定して整形表示ができる。

In [9]:
temperature = 23.4567
print(f"気温は {temperature} 度です。")
print(f"気温は {temperature:.2f} 度です。")

気温は 23.4567 度です。
気温は 23.46 度です。


| 書式指定       | 意味             | 例                    | 出力           |
| ---------- | -------------- | -------------------- | ------------ |
| `{x}`      | そのまま表示         | `{temperature}`      | 23.4567      |
| `{x:.1f}`  | 小数点以下1桁        | `{temperature:.1f}`  | 23.5         |
| `{x:.2f}`  | 小数点以下2桁        | `{temperature:.2f}`  | 23.46        |
| `{x:8.2f}` | 全体8桁・小数2桁（右寄せ） | `{temperature:8.2f}` | `'   23.46'` |


# 1-5. コメントアウトとエラーメッセージ

## なぜ型を意識する必要があるのか？

- **計算できるかどうかは型で決まる**  
    - たとえば、エクセルファイルで私たちの目に見える「5」という数字は、**文字列 `"5"`** の場合と **整数 `5`** の場合がある。  
    - 文字列のままでは計算に使えないため、一度 `int("5")` のように数値に変換する必要がある。

- 以下、# print(a+3) の行頭の # を除いた場合、どうなるか確認してみよう。
    - `#` はコメントアウトを表す記号で、この記号から始まる文章はプログラム実行時に無視される。
    - メモや説明文として利用できる他、エラーチェックの際に頻繁に用いる。

In [5]:
# 例：文字列と整数の違い
a = "5"   # 文字列
b = 5     # 整数

# 以下はエラーになる
# print(a + 3)

# 正しくは、文字列を整数に変換
print(int(a) + 3)

8


- 与えられた値が指定した型の変換に適していない場合、たとえば文字列を `int()` や `float()` に変換しようとするとエラーが発生する。

In [6]:
# OK
int("5")

# エラーになる例
int("five")   # ValueError: invalid literal for int()

ValueError: invalid literal for int() with base 10: 'five'

- 赤文字で表示された ValueError はエラーメッセージの1つで、与えられた値が変換に適していないことを指摘する。
- これらのエラーメッセージは、**プログラムのどこが間違っているのかを知る重要な手がかり**。はじめのうちは英語で書かれているため戸惑うかもしれないが、すぐに意味がわからない場合はChatGPTなどに貼り付けて解釈を確認するとよい。

- print 関数は文字列を + で結合して表示することもできる。
- print 関数では数値の代入された変数を引数としてその数値を出力することもできる。
- しかしながら、文字列と一緒に出力することはできない。この場合、str() を使って数値を文字列に変換することで、文字列同士を結合して出力できる。

In [7]:
print("今日の天気は"+"晴れです。")

今日の天気は晴れです。


In [1]:
temperature = 23.5
print(temperature)

23.5


In [2]:
temperature = 23.5
print("水温は " + temperature + " ℃です。")

TypeError: can only concatenate str (not "float") to str

In [ ]:
temperature = 23.5
print("水温は " + str(temperature) + " ℃です。")

# 1-6. 算術演算子と優先順位

- Python では数値を計算するために **算術演算子** を用いる。
- 四則演算のほか、累乗や余りの計算などもできる。多くはExcelで使用する演算子と同じ。
- 演算子には優先される順序があるため注意。

| 演算子  | 意味            | 例        | 結果  |
| ---- | ------------- | -------- | --- |
| `+`  | 加算            | `3 + 2`  | 5   |
| `-`  | 減算            | `5 - 1`  | 4   |
| `*`  | 乗算            | `2 * 4`  | 8   |
| `/`  | 除算            | `5 / 2`  | 2.5 |
| `//` | 整数除算（小数を切り捨て） | `5 // 2` | 2   |
| `%`  | 余り            | `5 % 2`  | 1   |
| `**` | 累乗            | `3 ** 2` | 9   |

- ⚠️ 演算子は ()  →  **  →  * / // %  →  + - の順で処理されるため注意。


In [4]:
a = 10 + 2 * 3
b = (10 + 2) * 3
c = 10 + 2 ** 3
print(a, b, c)

16 36 18


# 1-7. import文と標準ライブラリ

- Python にはあらかじめ多くの **標準ライブラリ（道具箱）** が用意されている。
- 必要なときに import 文を使って取り出して使う。たとえば、数学関数をまとめた「math」モジュールを使うと、円周率 π や平方根などを扱うことができる。

In [7]:
import math
# math.pi のピリオド . は "mathの中のpiを使う" という意味。
print(math.pi)       # 円周率 π
print(math.sqrt(16)) # 平方根

3.141592653589793
4.0


# Step 1の練習問題
1. "25" を整数に変換して +5 してみよう。

2. 10.2 を文字列に変換して "℃" と結合して出力せよ。

3. "True" を bool() に変換するとどうなるか試してみよう。また"" を bool() に変換した場合と比較してみよう。

4. 以下のプログラムのエラーを修正して実行せよ。

   print("The highest temperature in Los Angels in 2024 was" + 104 + " °F")

5. 次の変数が与えられている:

    station = "Tokyo Bay"
    temperature = 22.37
    salinity = 3.42
    DO = 8.1

    これらを使って、次のような1行出力を作成せよ。

    Area = Tokyo Bay, Temp = 22.37 ℃, Salinity = 3.42 %, DO = 8.10 mg/L

6. キーボードから今日の気温(摂氏)を入力し、華氏に変換させ、下記を表示させるプログラムを作成してみよう。CとFには入力値と変換値がそれぞれの値が代入されているものとする。

   今日の摂氏は C 度、華氏は F 度です。

   ※ 華氏と摂氏の関係は F = 1.8*C + 32

7. 5とは逆に、キーボードから今日の気温(華氏)を入力し、摂氏に変換させ、下記を表示させるプログラムを作成してみよう。CとFには入力値と変換値がそれぞれの値が代入されているものとする。

   今日の摂氏は C 度、華氏は F 度です。

8. 2023年の世界の二酸化炭素平均濃度は419.3 ppm、2024年の世界の二酸化炭素平均濃度は422.8 ppmであった。2023年から2024年にかけての二酸化炭素濃度の増加率は何%か計算し、下記を表示させるプログラムを作成してみよう。p にはパーセンテージとしての増加率が代入されているものとする。
  
   2023年から2024年にかけての二酸化炭素濃度の増加率は p %です。

9. 太陽放射の入射量を I = 1361 W/m² とし、地球の反射率（アルベド）を A = 0.3 とする。このとき、有効放射量（実際に地球が吸収する放射量）を求め、f-string整形で小数点以下2桁まで表示せよ。

    ※ 有効放射量 = (1 − A) × I
   
10.  地球の半径をr = 6371 kmとし、地球を球とみなして体積を概算せよ。この時、math.piを使用すること。また、最終出力はf-string整形で小数点以下2桁まで表示せよ。